src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala

```diff
@@ -1,48 +1,110 @@
 package org.broadinstitute.dsde.workbench.leonardo.api
 
+import akka.http.scaladsl.model.StatusCodes
 import akka.http.scaladsl.server.Directives._
-import akka.http.scaladsl.server.Route
+import akka.http.scaladsl.server.{AuthorizationFailedRejection, Directive0, Directive1, Route}
 import com.typesafe.scalalogging.LazyLogging
 import org.broadinstitute.dsde.workbench.leonardo.model.ClusterName
 import org.broadinstitute.dsde.workbench.leonardo.service.ProxyService
-import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
 import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+import akka.http.scaladsl.model.headers.{Authorization, HttpCookie}
+import org.broadinstitute.dsde.workbench.model.UserInfo
 
 import scala.concurrent.ExecutionContext
 
-/**
-  * Created by rtitle on 8/4/17.
-  */
-trait ProxyRoutes extends UserInfoDirectives{ self: LazyLogging =>
+trait ProxyRoutes extends UserInfoDirectives { self: LazyLogging =>
   val proxyService: ProxyService
   implicit val executionContext: ExecutionContext
 
   protected val tokenCookieName = "FCtoken"
 
   protected val proxyRoutes: Route =
+    path("notebooks" / "login") {
+      extractUserInfo { userInfo => // rejected with AuthorizationFailedRejection if the token is not present
+        setTokenCookie(userInfo) {  // set the token cookie in the response
+          complete {
+            logger.debug(s"Successfully logged in user $userInfo")
+            StatusCodes.OK
+          }
+        }
+      }
+    } ~
+    path("notebooks" / "logout") {
+      extractToken { token =>
+        complete {
+          proxyService.invalidateAccessToken(token).map { _ =>
+            logger.debug(s"Invalidated access token $token")
+            StatusCodes.OK
+          }
+        }
+      }
+    } ~
     path("notebooks" / Segment / Segment / "api" / "localize") { (googleProject, clusterName) =>
       extractRequest { request =>
-        cookie(tokenCookieName) { tokenCookie => // rejected with MissingCookieRejection if the cookie is not present
+        extractUserInfo { userInfo => // rejected with AuthorizationFailedRejection if the token is not present
           complete {
-            proxyService.getCachedUserInfoFromToken(tokenCookie.value).flatMap { userInfo =>
```

Comment:
`proxyService.getCachedUserInfoFromToken` has been lifted into the `extractUserInfo` directive.